ocaml/xapi/xapi_mgmt_iface.ml

```diff
@@ -178,20 +178,15 @@ let wait_for_management_ip ~__context =
   let is_connected () = Helpers.get_management_iface_is_connected ~__context in
   wait_for_ip get_ip is_connected
 
-let has_carrier ~__context ~self =
-  let metrics = Db.PIF.get_metrics ~__context ~self in
-  Db.PIF_metrics.get_carrier ~__context ~self:metrics
-
 (* CA-280237: Called in startup sequence after creating cluster_hosts *)
 let wait_for_clustering_ip ~__context ~(self : API.ref_Cluster_host) =
   let pIF = Db.Cluster_host.get_PIF ~__context ~self in
-  let iP = ref (Db.PIF.get_IP ~__context ~self:pIF) in
-  Mutex.execute ip_mutex (* Don't return until PIF is plugged AND has a valid IP *)
-    (fun () -> while !iP="" || not (has_carrier ~__context ~self:pIF) do
-        Condition.wait ip_cond ip_mutex;
-        iP := Db.PIF.get_IP ~__context ~self:pIF
-      done);
-  !iP
+  let network = Db.PIF.get_network ~__context ~self:pIF in
+  let interface = Db.Network.get_bridge ~__context ~self:network in
+  debug "Waiting for clustering IP on interface %s (%s)" interface (Ref.string_of pIF);
+  let get_ip () = Db.PIF.get_IP ~__context ~self:pIF in
```

Comment:
How about going a step further and doing a direct lookup of the IP as well? In the case of DHCP, the database will lag behind a little as well.

Reply:
There is a Helpers.get_primary_ip_addr, I'll use that.